airbyte-webapp/src/pages/source/SourceOverviewPage/sourceOverviewContext.ts

```diff
@@ -0,0 +1,25 @@
+import { useOutletContext } from "react-router-dom";
+
+import { SourceDefinitionRead, SourceRead, WebBackendConnectionListItem } from "core/request/AirbyteClient";
+import { useConnectionList } from "hooks/services/useConnectionHook";
+import { useGetSource } from "hooks/services/useSourceHook";
+import { useSourceDefinition } from "services/connector/SourceDefinitionService";
+
+interface SourceOverviewContext {
+  source: SourceRead;
+  sourceDefinition: SourceDefinitionRead;
+  connections: WebBackendConnectionListItem[];
+}
+
+export const useSourceOverviewContext = () => {
+  return useOutletContext<SourceOverviewContext>();
+};
+
+export const useSetupSourceOverviewContext: (id: string) => SourceOverviewContext = (id: string) => {
+  const source = useGetSource(id);
+  const sourceDefinition = useSourceDefinition(source.sourceDefinitionId);
+  // We load only connections attached to this source to be shown in the connections grid
+  const { connections } = useConnectionList({ sourceId: [source.sourceId] });
+
+  return { source, sourceDefinition, connections };
+};
```

Comment:
Since there's 3 items here referenced in 3 places, thought consolidating it would be useful.

Reply:
I would actually rather be in favor of just calling `useGetSource()`, `useSourceDefinition()` and `useConnectionList()` in the two routes (it's only 2 places, right? ). I think creating a hook just to avoid writing 3 lines in each route adds more indirection. It actually creates more code and you also have to follow `useSourceOverviewContext()` > `SourceItemPage` > `useSetupSourceOverviewContext()` to find the underlying API hooks. IMO this will make it harder to add or remove more data-fetching later.

Performance-wise I don't think there's much to gain here, right? `react-query` will cache responses and deduplicate requests for us, so we shouldn't worry about calling e.g. `useConnectionList()` too many times! If we see too many network requests, we should tweak our cache/stale/revalidation times in `react-query`.

Reply:
react-query doesn't actually cache the result. If you navigate to the pages today it doubles up on the request.
The intent of this ticket was to consolidate the requests in the higher level component so it's not fetched twice. The page movement and re-org was just a useful result of that goal. I can rename it or something, but I felt it was worth consolidating into two hooks so the data had a consistent contract.

Reply:
> react-query doesn't actually cache the result. If you navigate to the pages today it doubles up on the request.

I'm pretty sure it _does_ cache the result (or it should, that's kind of the main benefit of react-query) and you're just seeing duplicate requests because the `staleTime` is set to 0 by default. So the page is not suspended due to the second request, that just happens in the background. We could accomplish the same here by setting `staleTime` to something more reasonable (maybe 2 seconds) to avoid duplicate requests.

Reply:
I'm not sure I agree. I think this pattern gets the job done and it's an easy "Go-to Definition" click to see what's going on.
[DestinationItemPage](https://github.com/airbytehq/airbyte/blob/adc2c30ca3c6e9bb92c29ce6432c2644c8a5e1af/airbyte-webapp/src/pages/destination/DestinationItemPage/DestinationItemPage.tsx#L50) does this too, though with less data.

I think we may want to add a `staleTime` to the whole app. I assumed we had one, but that would be another ticket.

Reply:
Yep, both patterns definitely work to avoid extra network requests. The recent PR to do the same thing in [DestinationItemPage](https://github.com/airbytehq/airbyte/pull/22068) is the same deal - we can either work around calling `useGetDestination()` multiple times, or we could adjust the `staleTime` of that query. In my opinion we should work with `react-query` rather than around it - feel free to do as you see fit here, I just wanted to suggest an alternative solution 🙂

Reply:
I'd personally also feel that creating new context for just sharing responses of a query might be a bit overhead and causing us an "context inflation", while we have react-query, which can handle this for us, and make the consuming component looks nicer, because it's clearer that it gets data from an API. I think though this might be an overall discussion, and Joey wanted to present something abour react-query soon, which I hope we can use as a discussion for clarifying which way we'd prefer in the future.